### PR TITLE
Adjust path handling (+ tests) for Windows compatibility

### PR DESCRIPTION
**Pull request by colinrotherham**
Split out from https://github.com/alphagov/govuk-frontend/pull/2938

This PR makes various fixes for Windows compatibility

Our Windows support (slash direction) was partially prevented by:

~https://github.com/alphagov/govuk-frontend/blob/main/tasks/gulp/__tests__/after-build-package.test.js~
~https://github.com/alphagov/govuk-frontend/blob/main/tasks/gulp/__tests__/after-build-dist.test.js~

**Update:** I've used [`slash()`](https://www.npmjs.com/package/slash) to guarantee consistent copy/replace/rename changes on POSIX-style forward slashes but `path.normalise()` to put them back to Windows-style backslashes where needed—everything now works again

This PR is number 1) in a three-part special:

1. https://github.com/alphagov/govuk-frontend/pull/2946
2. https://github.com/alphagov/govuk-frontend/pull/2938
3. https://github.com/alphagov/govuk-frontend/pull/2945

### CODE REVIEW

File: package.json
Comment (colinrotherham) [inline package.json:26]: By stitching npm commands together with `&&` we saw difficulty detecting "exits"
https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#exit-codes-and-error-action-preference

Comment (colinrotherham) [reply package.json:26]: This also allows us to run npm scripts _without_ other scripts (like tests) also running

### Build + test
Includes npm `pre` and `post` scripts

```console
npm run build:compile
npm run build:package
npm run build:dist
```

### Build only (no tests)
Excludes npm `pre` and `post` scripts

```console
npm run build:compile
npm run build:package --ignore-scripts
npm run build:dist --ignore-scripts
```

File: lib/file-helper.js
Comment (colinrotherham) [inline lib/file-helper.js:39]: This tiny change means we can remove [`recursive-readdir`](https://www.npmjs.com/package/recursive-readdir) and our `getFilesByDirectory()` helper function

File: lib/file-helper.js
Comment (colinrotherham) [inline lib/file-helper.js:192]: Marked optional `[entries]` as a directory listing includes both files and directories

Only directories (not files) will have the `.entries` property available

```mjs
// Listing for components directory
const listing = await getListing(configPaths.components)

// Accordion directory entry
const accordion = listing.get('accordion')

// Accordion files
const accordionEntries = accordion.entries

// Accordion example files
accordionEntries.get('accordion.mjs')
accordionEntries.get('accordion.yaml')
accordionEntries.get('README.md')
```

File: package.json
Comment (romaricpascal) [reply package.json:26]: Clever solution to chaining commands 🙌🏻

**Review by romaricpascal — COMMENTED** [review]
That was quite the overhaul, but I like the new shape of after-build-package 🙌🏻 

Sorry for the number of comments, most are little things to help clarify what's going on. I think my only big concern is with us taking over the responsibility of recursively listing directories rather than leaving it to a tested package, though.

File: tasks/gulp/__tests__/after-build-dist.test.js
Comment (colinrotherham) [reply tasks/gulp/__tests__/after-build-dist.test.js:68]: @romaricpascal What action would you like to see for this issue?

File: tasks/gulp/__tests__/after-build-dist.test.js
Comment (colinrotherham) [reply tasks/gulp/__tests__/after-build-dist.test.js:37]: @romaricpascal Ahh I did have one in **file-helper.js** but I removed it 😆 

We've had review comments in the past to _not_ add extra helpers unless necessary

(Happy to make a new helper)

File: lib/file-helper.js
Comment (colinrotherham) [reply lib/file-helper.js:27]: Nice, you learn something every day 🙈

File: lib/file-helper.js
Comment (colinrotherham) [reply lib/file-helper.js:97]: Sounds good, still works 👍

File: tasks/gulp/__tests__/after-build-dist.test.js
Comment (colinrotherham) [reply tasks/gulp/__tests__/after-build-dist.test.js:129]: @romaricpascal I think I'll drop the commit as the old tests checks the file could be read + truthy contents

File: tasks/gulp/__tests__/after-build-package.test.js
Comment (colinrotherham) [reply tasks/gulp/__tests__/after-build-package.test.js:70]: @romaricpascal Would be brill but I'm happy with like-for-like for now, can enhancements come in another PR?

Do flag regressions though—things that the old implementation did that I've broken

File: tasks/gulp/__tests__/after-build-dist.test.js
Comment (colinrotherham) [reply tasks/gulp/__tests__/after-build-dist.test.js:68]: Easier if I drop this commit, I'll leave the old tests as-is

File: tasks/gulp/__tests__/after-build-dist.test.js
Comment (colinrotherham) [reply tasks/gulp/__tests__/after-build-dist.test.js:101]: Going to leave the old tests in place

File: tasks/gulp/__tests__/after-build-dist.test.js
Comment (colinrotherham) [reply tasks/gulp/__tests__/after-build-dist.test.js:112]: Sadly this is coming out, unrelated but yeah I did wonder!

File: tasks/gulp/__tests__/after-build-package.test.js
Comment (colinrotherham) [reply tasks/gulp/__tests__/after-build-package.test.js:47]: In JavaScript "filter" means "keep it in the array" so I'm going to avoid **Filter out** as really it should be **Filter in** and then "shouldn't" flips it again 😆 

I'll keep it generic:

```suggestion
      // Only include files matching filter patterns
```

File: tasks/gulp/__tests__/after-build-package.test.js
Comment (colinrotherham) [reply tasks/gulp/__tests__/after-build-package.test.js:51]: I've gone with:

```suggestion
      // Replaces source component '*.mjs' with:
      // - `*.mjs` file copied to `govuk-esm`
      // - `*.js` file compiled to `govuk`
```

File: tasks/gulp/__tests__/after-build-package.test.js
Comment (colinrotherham) [reply tasks/gulp/__tests__/after-build-package.test.js:68]: I've gone with

```suggestion
      // Replaces source component '*.yaml' with:
      // - `fixtures.json` fixtures for tests
      // - `macro-options.json` component options
```

File: tasks/gulp/__tests__/after-build-package.test.js
Comment (colinrotherham) [reply tasks/gulp/__tests__/after-build-package.test.js]: Helper now in 😊

File: tasks/gulp/__tests__/after-build-dist.test.js
Comment (colinrotherham) [reply tasks/gulp/__tests__/after-build-dist.test.js]: With us dropping one of the commits I've moved this back up, copying the what we do in [**asset-version.js**](https://github.com/alphagov/govuk-frontend/blob/tests-for-windows/tasks/asset-version.js#L11)

```mjs
const pkg = require(join(cwd(), configPaths.package, 'package.json'))
```

File: tasks/gulp/__tests__/after-build-dist.test.js
Comment (colinrotherham) [reply tasks/gulp/__tests__/after-build-dist.test.js]: Two new helpers added for filtering and/or mapping: https://github.com/alphagov/govuk-frontend/pull/2946/commits/0d75191e5ad156204db3ff036b75f2e27b8a707d

File: tasks/gulp/__tests__/after-build-dist.test.js
Comment (colinrotherham) [reply tasks/gulp/__tests__/after-build-dist.test.js]: Two new helpers added for filtering and/or mapping: https://github.com/alphagov/govuk-frontend/pull/2946/commits/0d75191e5ad156204db3ff036b75f2e27b8a707d

File: tasks/gulp/__tests__/after-build-package.test.js
Comment (colinrotherham) [reply tasks/gulp/__tests__/after-build-package.test.js]: Two new helpers added for filtering and/or mapping: https://github.com/alphagov/govuk-frontend/pull/2946/commits/0d75191e5ad156204db3ff036b75f2e27b8a707d

File: tasks/gulp/__tests__/after-build-package.test.js
Comment (colinrotherham) [reply tasks/gulp/__tests__/after-build-package.test.js]: Refactored along with those two new helpers: https://github.com/alphagov/govuk-frontend/pull/2946/commits/0d75191e5ad156204db3ff036b75f2e27b8a707d

File: postcss.config.js
Comment (colinrotherham) [reply postcss.config.js:19]: Done 👍 

```suggestion
  const { dir, name } = parse(
    // Normalise to string path with forward slashes
    slash(typeof file === 'object' ? file.path : file)
  )
```

File: tasks/gulp/__tests__/after-build-package.test.js
Comment (romaricpascal) [reply tasks/gulp/__tests__/after-build-package.test.js]: Yeah, I think it'd be fine to have it in a separate PR so that this PR is like-for-like, that's a good call 😄

File: tasks/gulp/__tests__/after-build-package.test.js
Comment (romaricpascal) [reply tasks/gulp/__tests__/after-build-package.test.js]: If we want to avoid "Filter", could we go for "Exclude files in the source listing that shouldn't be in the package"?

A generic comment like that last suggestion doesn't bring much to the table when you read the file, as this is just a translation of `.filter(filterPath(filterPatterns))`, and with the addition of the helper, it'll already kind of clear. 

What's less clear and I wanted to highlight in the comment is the role of that computation in checking the package, which is to weed out from the expected files those that we know won't be in the package for sure.

Sorry if it sounds nitpicky, but I think having such a trail will help the next person reading that file 😄

File: postcss.config.js
Comment (romaricpascal) [reply postcss.config.js]: Similarly to the filtering comment, I was hoping we could highlight why we need to normalise the slashes to being forward slashes, as you may expect that to be already handled by `parse`, while actually the reason is for use with the `dir.endsWith` further down the line.

File: tasks/gulp/__tests__/after-build-package.test.js
Comment (romaricpascal) [inline tasks/gulp/__tests__/after-build-package.test.js:86]: **issue**: At second read, the listing for the actual files shouldn't exclude the same filterPatterns as the expected files. By doing it in both places, we'll miss if one of the excluded files has been included by error.

File: lib/file-helper.js
Comment (romaricpascal) [reply lib/file-helper.js:40]: I'm very torn on that abstraction, sorry 😅 . It's technically a nice structure to represent filesystem nodes. But I'm not sure it's necessary when `glob` or `recursive-readdir` would give us a flat list of files over which to `filter`, `flatMap` in our tests:
- already tested and documented out of the box
- in a straighforward/well-known syntax, without having to `flatMap` with `listingToArray` first

The tested bit feels quite important to me there, as it's ultimately code that supports our own tests (not that a 3rd party package will be bug-proof, but they'll have tests around what that specific package does, while our helpers are more tested through our tests).

It feels to me like it makes things a bit more complex than they should be, but maybe it's just me 😊  I think I'd welcome a second opinion from someone in @alphagov/design-system-developers on that one.

Comment (colinrotherham) [reply lib/file-helper.js]: @romaricpascal You can sleep well tonight 😊 

All directory listings have now been swapped with `glob()` everywhere

Ready for review again

File: lib/file-helper.js
Comment (colinrotherham) [reply lib/file-helper.js]: Swapped with `glob()` as discussed in https://github.com/alphagov/govuk-frontend/pull/2946#discussion_r1007203034

File: tasks/gulp/__tests__/after-build-package.test.js
Comment (colinrotherham) [reply tasks/gulp/__tests__/after-build-package.test.js]: @romaricpascal I've removed the comment (saves this being a blocker)

If there's something in the PR that's less understandable than it was before let me know 🙈

File: postcss.config.js
Comment (colinrotherham) [reply postcss.config.js]: I've not picked this one up, is it a blocker?

File: tasks/gulp/__tests__/after-build-package.test.js
Comment (colinrotherham) [reply tasks/gulp/__tests__/after-build-package.test.js]: @romaricpascal This was a great spot, thanks

It was slightly useful at ignoring `.DS_Store` file if you'd opened the destination directories in Finder, but yep we don't want any stray tests or snapshots being wrongly ignored

All filters have now been removed from the "expected" arrays

File: gulpfile.js
Comment (romaricpascal) [inline gulpfile.js:44]: **question**: As we use `join`, won't Gulp and Node sort themselves out without us having to `slash` regarding Windows vs. Posix paths?

File: lib/file-helper.js
Comment (romaricpascal) [inline lib/file-helper.js:26]: Thanks for making that update and going through all the changes for it  🙌🏻 🙌🏻 🙌🏻

File: src/govuk/components/components.test.js
Comment (romaricpascal) [inline src/govuk/components/components.test.js:16]: I love how clear that is 😍

File: src/govuk/helpers/helpers.test.js
Comment (romaricpascal) [inline src/govuk/helpers/helpers.test.js:33]: *question*: This checks the shape of the return of `renderSass`, which we can trust Sass' `.render` to do properly. As [the method throws a `LegacyException` if the compilation fail](https://sass-lang.com/documentation/js-api/modules#render), and in turn fail `Promise.all`, wouldn't it be enough to no check what's returned and have that a test that fails on exception rather than unmet expectations (which will always be met)?

File: src/govuk/helpers/helpers.test.js
Comment (romaricpascal) [inline src/govuk/helpers/helpers.test.js:55]: Cheers for the little tidy up 🙌🏻

File: tasks/gulp/__tests__/after-build-package.test.js
Comment (romaricpascal) [inline tasks/gulp/__tests__/after-build-package.test.js:70]: *question*: Why are these strings normalised while there's no slashes in them?

File: tasks/gulp/__tests__/after-build-package.test.js
Comment (romaricpascal) [inline tasks/gulp/__tests__/after-build-package.test.js:22]: We may have a slight naming issue there, as I thought `configPaths.src` was the `src` folder, while it's actually `src/govuk`. 

*suggestion*: I think a little comment will help the next person reading that, (or maybe `join(cwd(), 'src')` to get the path), as a quick way out.

```suggestion
    // Join with `../` to get the content of the `src` directory
    listingSource = await getListing(join(configPaths.src, '../'))
```

**Review by romaricpascal — COMMENTED** [review]
So nice, great work! Looks all neat and feels a bit easier to follow with the use of globs and explicit `join` in the code. Thanks for making the update 🙌🏻 

Only had a few questions for clarification (slashing with Gulp's `dest` and new expectations on the output of renderSass, mainly) 😄

File: gulpfile.js
Comment (colinrotherham) [reply gulpfile.js:44]: Sorry @romaricpascal there's a note in the commit message:

>Note: Gulp globs use `slash()` to ensure Windows paths are flipped back to POSIX for compatibility

It's because [Gulp uses its own glob library](https://github.com/gulpjs/glob-parent#windows) and things (can) go wrong in strange ways. Sadly too `require('path/posix').normalize` doesn't flip Windows slashes either

File: src/govuk/helpers/helpers.test.js
Comment (colinrotherham) [reply src/govuk/helpers/helpers.test.js:33]: @romaricpascal Probably, I'm just matching these tests with the same format we use here:
https://github.com/alphagov/govuk-frontend/blob/main/src/govuk/components/components.test.js#L29

We can change them all in another PR if it's a concern?

File: tasks/gulp/__tests__/after-build-package.test.js
Comment (colinrotherham) [reply tasks/gulp/__tests__/after-build-package.test.js:70]: @romaricpascal Good spot, we use `cwd` in **glob()** to use relative paths so I'll remove these

(Previously the paths were `package/` suffixed)

File: src/govuk/helpers/helpers.test.js
Comment (romaricpascal) [reply src/govuk/helpers/helpers.test.js:33]: Ah, I didn't see that file had that structure for this test. Happy with aligning them all to do the same then 😄

File: gulpfile.js
Comment (romaricpascal) [reply gulpfile.js:44]: Ok, cool. I'm not sure if we'd use globs on `gulp.dest` but better safe than sorry.

**Review by romaricpascal — APPROVED** [review]
Approving so it can get merged after the `normalize` calls are removed 😄 🎉 🎉 🎉

File: gulpfile.js
Comment (colinrotherham) [reply gulpfile.js:44]: Yeah, definitely. Also affects non-glob things like `./` and `../` in joined paths too

**Pull request merged**
